cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceService.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.namespace;
+
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.app.store.StoreFactory;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.exception.AlreadyExistsException;
+import co.cask.cdap.common.exception.NotFoundException;
+import co.cask.cdap.config.DashboardStore;
+import co.cask.cdap.config.PreferencesStore;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.NamespaceMeta;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link AbstractIdleService} for managing namespaces
+ */
+public final class NamespaceService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(NamespaceService.class);
+  private static final String NAMESPACE_ELEMENT_TYPE = "Namespace";
+
+  private volatile Store store;
+  private final StoreFactory storeFactory;
+  private final PreferencesStore preferencesStore;
+  private final DashboardStore dashboardStore;
+
+  @Inject
+  public NamespaceService(StoreFactory storeFactory, PreferencesStore preferencesStore, DashboardStore dashboardStore) {
+    this.storeFactory = storeFactory;
+    this.preferencesStore = preferencesStore;
+    this.dashboardStore = dashboardStore;
+  }
+
+  /**
+   * Initializes and starts namespace service.
+   * Currently only creates the default namespace if it does not exist.
+   * In future, it could potentially do other tasks as well.
+   */
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting namespace service");
+    initializeStore();
+    ensureDefaultNamespaceExists();
+  }
+
+  /**
+   * Asynchronously initializes the #store, with retries.
+   * Need a retry mechanism since {@link StoreFactory#create()} requires the dataset service to be running.
+   */
+  private void initializeStore() {
+    new RetryThread(new Runnable() {
+      @Override
+      public void run() {
+        store = storeFactory.create();
+      }
+    }, 2, TimeUnit.SECONDS).run();
```

Comment:
So by calling `run()`, it's not async. You are simply executing the Runnable directly from this thread. Is it what you want? Seems unmatch with the javadoc.